src/Analyzers/CSharp/Analyzers/UseSystemThreadingLock/CSharpUseSystemThreadingLockDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,235 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseSystemThreadingLock;
+
+/// <summary>
+/// Looks for code of the form:
+/// 
+///     private ... object _gate = new object();
+///     ...
+///     lock (_gate)
+///     {
+///     }
+///     
+/// and converts it to:
+/// 
+///     private ... Lock _gate = new Lock();
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal sealed class CSharpUseSystemThreadingLockDiagnosticAnalyzer()
+    : AbstractBuiltInCodeStyleDiagnosticAnalyzer(
+        IDEDiagnosticIds.UseSystemThreadingLockDiagnosticId,
+        EnforceOnBuildValues.UseSystemThreadingLock,
+        CSharpCodeStyleOptions.PreferSystemThreadingLock,
+        new LocalizableResourceString(
+            nameof(CSharpAnalyzersResources.Use_System_Threading_Lock), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+{
+    /// <summary>
+    /// A method body edit anywhere in a type will force us to reanalyze the whole type.
+    /// </summary>
+    /// <returns></returns>
+    public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+    protected override void InitializeWorker(AnalysisContext context)
+    {
+        context.RegisterCompilationStartAction(compilationContext =>
+        {
+            var compilation = compilationContext.Compilation;
+
+            // The new 'Lock' feature is only supported in C# 13 and above, and only if we actually have a definition of
+            // System.Threading.Lock available.
+            if (!compilation.LanguageVersion().IsCSharp13OrAbove())
+                return;
+
+            var lockType = compilation.GetTypeByMetadataName("System.Threading.Lock");
```

Comment:
This is going to break if any metadata reference targets an earlier version of .NET and defines its own System.Threading.Lock as an internal type.

Reply:
Yes.  But if you're on C#13, and somehow referencing such a thing... then that's clearly pathological.  :)

Reply:
updated to at least make sure the type is public.

Reply:
> Yes. But if you're on C#13, and somehow referencing such a thing... then that's clearly pathological. :)

This can be caused by _any_ reference doing this. For example, you could be on C# 13 and .NET 8, but reference a nestandard2.0 library that has an internal definition of System.Threading.Lock that otherwise has no impact on your project. But it will break this analyzer.

Reply:
> that has an internal definition of System.Threading.Lock that otherwise has no impact on your project. But it will break this analyzer. 

The code won't accept an internal definition now.

Reply:
Added test demonstrating this.  You'll need to be on C#13, referencing an assembly that has a public System.Threading.Lock, with a public inner type called 'Scope'.  If you have that, and you're not a Lock, that's on you.  because that's all the compiler checks for as well :)